### PR TITLE
Implement uniform floats with decreasing distance towards 0.0

### PR DESCRIPTION
**Pull request by RaimoNiskanen**
This pull request implements `rand:uniform_nz/0` and `rand:uniform_nz_s/1`.

These new functions return an uniformly distributed random float in the range ]0.0;1.0] that is the half open unity interval not including zero but including one.  The interesting property of this implementation is that for all generated floats all bits are random regardless of the exponent of the float, and they are still uniformly distributed.

The existing function `rand:uniform/0` generates values on the form N * 2^-53 so they all have the same distance to the closest possible value.  But with this function that distance decreases towards 0.0 as does the probability for a specific value, hence uniformity is preserved.

This is done by iteratively generating more bits until there are 53 random bits counting from the generated top `1` while keeping track of the resulting exponent.

This kind of uniformity could maybe be classified as "geometrical" as opposed to the existing "arithmetical" uniformity.  It should useful not simply when you do not want the number 0.0, but when you will calculate e.g 1/X or log(X) of the generated number since the "arithmetical" distribution will be wildly scattered by those functions.  An example of this is when creating normal distributed values using the Box-Muller transformation that calculates log(X) on one number and sin(Y), cos(Y) on a second number.  Here a "geometrical" distribution should be appropriate for the first number while an "arithmetical" distribution should be appropriate for the second.
 
The execution speed as indicated by `rand_SUITE:measure/1` is about half that of the existing float generator, which I guess is the price you might have to pay for this property.  It could maybe, with some effort be optimized to about 50% to 70% slower than the existing.

### CODE REVIEW

**Comment by okeuday** [comment]
@RaimoNiskanen I don't think this function should be slower.  An example of both is at https://github.com/okeuday/quickrand/blob/1c60daa4b0edcec0a6b8310443574baac962ea88/src/quickrand.erl#L328-L330 (for `uniform`) and https://github.com/okeuday/quickrand/blob/1c60daa4b0edcec0a6b8310443574baac962ea88/src/quickrand.erl#L363-L365 (for `uniform_nz`).

**Comment by RaimoNiskanen** [comment]
@okeuday I think you are comparing apples with pears. A uniform generator with _arithmetic_ uniformity distribution but left open interval does of course not have to be slower than a for a right open interval.  And such a generator is easily implemented with `1.0 - rand:uniform()`.

I think that _geometric_ uniformity distribution would fit like a glove for the left open interval since that uniformity distribution implicates that zero should not be returned because the possibility is infinitely low.

**Comment by okeuday** [comment]
@RaimoNiskanen Yes, when I saw the `uniform` name prefix on this function I was hoping it represented a uniform distribution, but it doesn't.  Why is it misleading?

**Comment by RaimoNiskanen** [comment]
It is not misleading.  The distribution is uniform, but the spacing between the numbers is not.

The existing `rand:uniform()` has got equal spacing between the generated numbers. The generated numbers are on the form `N * 2^-53` where N is an integer [0;2^53}

The proposed `rand:uniform_nz()` can generate any float that the IEEE format (normalized) allows, in the interval (0;1] and the spacing between the possible IEEE numbers decreases towards zero.

And the distribution is uniform.

**Comment by okeuday** [comment]
@RaimoNiskanen The distribution is not uniform because the spacing between number is not uniform.  If you were to do a scatter plot of the result of some large number of samples, a uniform plot is a horizontal line (with y axis as float value count and x axis as float value).  The function you have described as `uniform_nz` is best described as "skew" because it will be a somewhat diagonal line, due to this spacing between numbers not being uniform (the random distribution of values coming from `uniform_nz` will not be uniform, contradicting its name).

**Comment by RaimoNiskanen** [comment]
The distribution from `uniform_nz` is uniform since for every right open sub-range in the interval (0;1] the probability to draw a number in any other sub-range of equal size is the same.  This does _not_ mean that every possible number is equally probable.

A scatter plot is not good to see uniformity.  Use a histogram.

See http://xoroshiro.di.unimi.it/ the section "Generating uniform doubles in the unit interval"; the last paragraph.

**Comment by okeuday** [comment]
@RaimoNiskanen The normal interpretation of uniform distribution is "every possible number is equally probable" (in the whole range (0..1]) and I do not see how you can call this distribution uniform.  I could say a sub range of the tails on a normal curve looks horizontal and uniform, but it isn't a uniform distribution.  The reason I complain about this function, is because I see it as misleading.  I consider the uniform implementation as:

    -define(DBL_EPSILON_DIV2, 1.1102230246251565e-16).
    ...
    <<I:53/unsigned-integer, _:3>> = crypto:strong_rand_bytes(7),
    (I + 1) * ?DBL_EPSILON_DIV2.

I struggle to see why you should use a slower function with a distribution that is not as uniform.  I understand your solution for the code above would be `1.0 - rand:uniform()` to get (0..1], and is why you have claimed `rand:uniform_nz` should not be implemented with code similar to what is above, but I don't see an advantage to the slower alternative in this pull request that lacks uniformity, except to confuse developers.

**Comment by sverker** [comment]
There is a difference between
https://en.wikipedia.org/wiki/Discrete_uniform_distribution
and
https://en.wikipedia.org/wiki/Uniform_distribution_(continuous)

and as I see it, `uniform_nz` would be the closest approximation to a continuous uniform distribution in (0..1] that is possible with 64-bit floats.

**Comment by okeuday** [comment]
@RaimoNiskanen @sverker Ok, based on information at http://xoroshiro.di.unimi.it/random_real.c I would assume that `uniform_nz` is attempting to be a more exact uniform distribution that handles a nonuniform distribution of floating point values due to the IEEE double precision floating point format, while `rand:uniform/0` represents an inexact uniform distribution because of double precision rounding skewing the distribution (though that is minimized by choosing 53 bits of randomness, instead of something like 64 bits).

If that is correct, then that should also mean that doing `1.0 - rand:uniform_nz()` would create a distribution in [0..1) that is not exactly uniform due to this same rounding problem.  That would mean that `rand:uniform()` and `1.0 - rand:uniform_nz()` do not provide an exact distribution in [0..1) and another function for this would be necessary, if it was needed.  Also, due to this complexity, it seems best to differentiate the `rand:uniform/0` function from the `rand:uniform_nz/0` function with a suffix, like `uniform_exact_nz` which then allows the addition of other functions in the future, if that is necessary.  Having a note in the documentation to make this clearer would help a lot, because it could say "yes this is slower, but this is more accurate" which helps to give the developer some indication of how it would be applied to source code.

**Comment by RaimoNiskanen** [comment]
@okeuday That is about right. `rand:uniform_nz()` provides uniform distribution of all possible IEEE double precision floating point numbers.  Not nonuniform!  As @sverker said, the closest approximation to a continuous *uniform* distribution possible with IEEE floats.

I tried to document this but have not received any comments on the documentation.  The wording about "closest approximation of continuous distribution" should probably be added, though.  That it would be more accurate is a matter of context, though.

Using `1.0 - rand:uniform_nz()` would cause rounding problems.

`rand:uniform/0` does *not* give an inexact uniform distribution and there is no rounding skewing *because* it uses 53 bits of randomness.   The generated numbers are actually a discrete uniform distribution translated to IEEE floating point numbers as `N * 2^-53` with 0 =< integer(N) < 2^53.  So all numbers it can generate are equally probable.  Furthermore `1.0 - rand:uniform()` will *because* of the set of numbers the function generates *not* introduce any rounding.  It is an *exact* operation for all generated numbers.  The resulting numbers will be on the same form `N * 2^-53` with 0 < integer(N) =< 2^53.

I do not know if the infix `exact` is "right".  But it is true that the "nonzero" property is kind of a consequence of its notion of uniformity.

**Comment by sverker** [comment]
Without giving it too much detailed thought, I would think `1.0 - rand:uniform_nz()` yielding about the same distribution as `uniform/0`.
Values in (0.5, 1] will map to [0, 0.5) and keep their 2^-53 resolution.
Values in (0, 0.5] will map to [0.5, 1) and be rounded to nearest N*2^-53.

**Comment by okeuday** [comment]
@RaimoNiskanen @sverker There is a logical problem here:  If `rand:uniform/0` is exactly uniform with its IEEE double precision result, then there should be no reason for `rand:uniform_nz/0` to use a slower algorithm.  A condensed example is below:

    -define(DBL_EPSILON_DIV2, 1.1102230246251565e-16).
    <<I:53/unsigned-integer, _:3>> = crypto:strong_rand_bytes(7),
    I * ?DBL_EPSILON_DIV2.

The above code is similar to how `rand:uniform/0` makes a double precision floating point value in [0..1) while this can be adjusted for the range (0..1], as shown below:

    <<I:53/unsigned-integer, _:3>> = crypto:strong_rand_bytes(7),
    (I + 1) * ?DBL_EPSILON_DIV2.

The two are very similar and use overlapping integers, with the only difference affecting (in the second block of source code) the result of 1.0, which should not be a concern for rounding because it is part of the range required, (0..1].

If `rand:uniform/0` was exact, there should be no reason to use a more complicated and slower approach in `rand:uniform_nz/0`, because similar source code handles the (0..1] range you want too.

**Comment by RaimoNiskanen** [comment]
@sverker That is right - it will be about the same distribution, but the rounding to nearest N*2^-53 when mapping from (0, 0.5] to [0.5, 1) will per default (Gnu libc) be rounded to even so every second number will be more probable, in particular for (0.25, 0.5] to [0.5, 0.75) where every second number in the source range will be rounded, so that may mean (have not worked this through) that every second number in the destination range will get 3 times the probability of the numbers in between.

But, about the same distribution, yes.

**Comment by RaimoNiskanen** [comment]
`rand:uniform_nz` is intended for applications that benefit from higher granularity near zero for example `(1 / rand:uniform_nz())` or `math:log(rand:uniform_nz())`.  The latter is used in the Box-Muller transformation for normal distribution.

When using `(1.0 - rand:uniform())` in the Box-Muller transformation the largest possible deviation will be about 8.6 with the closest neighbor about 1% away (less than 7 effective random bits).

When using `rand:uniform()` instead the largest possible deviation will be about 38.  The closest neighbor at 8.6 will be about a factor 10^(-15) away (roughly 50 effective random bits).

By the way: what do you mean with:
> If rand:uniform/0 was exact ...

Do you implicate that it is not exact?

**Comment by RaimoNiskanen** [comment]
I have a name suggestion.  What about `rand:uniform_real()` to indicate that it approximates an uniform distribution of random real numbers.

I have just realized that the interval ]0.0, 1.0] is not the appropriate value interval for `rand:uniform_real()`.

It should rather be [0.0, 1.0[, but not exactly 0.0.  Since the probability for exactly 0.0 is extremely low we might as well guarantee that it will never be returned.

This is because all other aspects except not returning exactly 0.0 point to a right open interval.

For example, for all sub ranges [0.0, 2^(-N)[ the probability is 2^(-N), and for all sub ranges [N * 2^(-K), (N + 1)*2^(-K)[ the probability is 2^(-K).  This works only with right open intervals, and also applies to `rand:uniform()` that has got a right open interval.

When implementing a half open interval it turns out to be a significantly smaller speed penalty.  `rand:uniform_real()` will only be about 25% slower than `rand:uniform()`.  It has to be slower since it amortized has to generate more bits to fill the mantissa.

**Comment by okeuday** [comment]
@RaimoNiskanen Just trying to understand why `rand:uniform_nz/0` is choosing to use a slower algorithm to get a result that should be more accurate (since you could also choose the faster way, mentioned above, which is closer to how `rand:uniform/0` works).  Part of the current problem is function naming where `rand:uniform/0` and `rand:uniform_nz/0` look like they should be using a similar algorithm, but they aren't.  Currently, only the "_nz" suffix allows them to differ, and we know this doesn't refer to New Zealand, so it must be Non-Zero.  Changing this suffix to "_slow_nz" would be more descriptive if you don't like "_exact_nz" (nothing here is really exact due to it being double precision).

It sounds like you want to add another function to provide the `rand:uniform/0` range of [0..1) using the slower algorithm that `rand:uniform_nz/0` is using.  Ideally, the common algorithm should share a common function name prefix to keep the common functionality clearly visible at a glance.

**Comment by RaimoNiskanen** [comment]
@okeuday So I guess you did not read my suggestion to rename the function to `uniform_real`?

I think `_slow` is a dreadful suffix/infix since the purpose is not to be slow, and 25% is not that slow for the increased granularity you get.

**Comment by okeuday** [comment]
@RaimoNiskanen Yes, "_real" is a good suffix, I was distracted by the range details and I still find it important to have "_nz" to show it is non-zero.  So, I would rather have `uniform_real_nz` which allows the range [0..1) to be used by `uniform_real` if it is added, though it is unclear [based on your recent comment](https://github.com/erlang/otp/pull/1574#issuecomment-333076945) if the function in this pull request is really (0..1].  Either way, a name of that form will help make its difference from `rand:uniform/0` clearer, along with documentation details.

File: lib/stdlib/doc/src/rand.xml [hunk 1]
```diff
@@ -478,6 +523,68 @@ end.</pre>
     </func>
 
     <func>
+      <name name="uniform_real_s" arity="1"/>
```
Comment (RaimoNiskanen) [inline lib/stdlib/doc/src/rand.xml]: @okeuday This section tries to explain the properties and the value range of `uniform_real_s/1`.  As you see it is essentially [0.0,1.0) but not exactly 0.0, and I find it tricky to explain.

**Comment by RaimoNiskanen** [comment]
@okeuday The "not returning zero" property of `rand_real` is a consequence of the value range.  I could remove that guarantee and say that the value range is simply [0.0,1.0).  Then the "truncate to closest IEEE 754 number" algorithm leads to the probability to get exactly 0.0 is about 2^-1022 i.e 2.2e-308.  A dirty calculation gives that the probability for that to happen before the sun burns out if you generate one number per nanosecond is still about 1e-282.

So, the value range is [0.0,1.0) but the function will not return exactly 0.0 in anybody's lifetime, therefore we might as well guarantee that to soothe programmers using it.  The same programmers that rely on that a hash collision in SHA1 will not occur (Git users, probability about 1e-24), and ignore the probability for an unrecoverable hard drive bit error (about 1e-14).  These errors are *hundreds of orders of magnitude* more probable...

The problem is to describe this in a quick and easy way.

And I think there is no room for a function based around approximating real numbers that has the value range (0.0,1.0] because of how IEEE 754 numbers work.  To include 1.0 gives automatically a skewed probability for that value range. Excluding 0.0 raises the difficult question of how close to 0.0 that should be excluded.  Therefore I think naming the function `uniform_real_nz` is redundant since not returning zero comes with `_real`.  This can not be described by just the function's name anyway.

**Comment by okeuday** [comment]
@RaimoNiskanen I understand the probability of 0.0 is low, however, because it is so much easier to deal with guarantees than very small probabilities, why not change this slightly to make the range (0..1) (so neither 0 nor 1 will occur, ever, not in anyone's lifetime).  With the algorithm you are using, but modified to be similar to:

    strong_floatM() ->
        <<I:53/unsigned-integer, _:3>> = crypto:strong_rand_bytes(7),
        if
            I == 0 ->
                % almost never executes this case, an additional function call
                % is necessary to have a uniform distribution
                strong_floatM();
            true ->
                I * ?DBL_EPSILON_DIV2
        end.

I agree the "_nz" suffix isn't as helpful for this function, and it would be a simpler situation without the suffix and the potential for having to think about a remote 0.0 possibility.

**Comment by RaimoNiskanen** [comment]
@okeuday So, have you looked at the documentation I suggested. There I say the interval is [DBL_MIN,1.0) and that one might as well regard that as "[0.0,1.0) but not 0.0", which may be regarded as (0.0,1.0).

**Comment by RaimoNiskanen** [comment]
Oh, and the implementation actually does as you suggests i.e to retry for exactly 0.0, just to prove it will never happen.  A lost clause...

**Comment by okeuday** [comment]
@RaimoNiskanen Sounds good.  The documentation has a clear interval (the comments above have been a bit tangential) and describes the basic difference with `rand:uniform/0`.

**Comment by RaimoNiskanen** [comment]
I squashed irrelevant commits

**Pull request merged**